src/main/java/io/cdap/plugin/mainframe/MainframeRecordReader.java

```
@@ -0,0 +1,307 @@
+
+    String recordDeciderFieldname = config.getRecordDeciderFieldname();
+    if (recordDeciderFieldname != null && !recordDeciderFieldname.trim().isEmpty()) {
+      CobolInputFormat.setRecordDeciderField(job, config.getRecordDeciderFieldname());
+      CobolInputFormat.setRecordDeciderCondition(job, config.getRecordDeciderConditions());
+    }
+
+    String recordSelectorFieldMap = config.getRecordSelectorFieldMappings();
+    if (recordSelectorFieldMap != null && !recordSelectorFieldMap.trim().isEmpty()) {
+      CobolInputFormat.setRecordSelectorFieldMaps(job, config.getRecordSelectorFieldMappings());
+    }
+
+    String recordSelectorExpression = config.getRecordSelectorExpression();
+    if (recordSelectorExpression != null && !recordSelectorExpression.trim().isEmpty()) {
+      CobolInputFormat.setRecordSelectorExpression(job, config.getRecordSelectorExpression());
+    }
+
+    context.setInput(Input.of(
+      config.referenceName, new SourceInputFormatProvider(
+        CobolInputFormat.class, job.getConfiguration()
+      ))
+    );
+
+    // Record lineage for the fields.
+    // Please note: Lineage implementation doesn't include
+    try {
+      LayoutDetail layout = config.getCopybookLayout();
+      MainframeRecordConvertors convertors = new MainframeRecordConvertors(layout);
+      Schema schema = convertors.toSchema();
+      LineageRecorder lineageRecorder = new LineageRecorder(context, config.referenceName);
+      lineageRecorder.createExternalDataset(schema);
+      StringBuilder sb = new StringBuilder("Reading cobol data from '");
+      sb.append(config.getInputPath()).append("', based on the provided copybook. ");
+      lineageRecorder.recordRead(
+        "Read",
+        sb.toString(),
+        Preconditions.checkNotNull(schema.getFields())
+          .stream()
+        .map(Schema.Field::getName)
+        .collect(Collectors.toList())
+      );
+    } catch (IOException | MainframeFieldTypeMappingException e) {
```

if it is an IOException, can we throw it as is, instead of wrapping it inside another IOException?

---

We are throwing as is. If we don't catch and rethrow errors are hidden way down in stack.

---

done